docs/blog/2019-03-03-new-schema-customization/index.md

```diff
@@ -0,0 +1,269 @@
+
+```js
+const Foo = graphql.GraphQLObjectType({
+  name: "Foo",
+  fields: () => ({
+    id: {
+      type: new graphql.GraphQLNonNull(graphql.GraphQLID),
+    },
+    bar: {
+      // Must be an actual reference to that type
+      type: Bar,
+    },
+  }),
+})
+```
+
+To solve this issues, a pattern called _Type Registry_ has been widely used. Type registry is some kind of abstraction that holds types inside it and allows other types to retrieve them.
+
+```js
+// some global state type registry
+const TypeRegistry = require("./typeRegistry")
+
+const Foo = graphql.GraphQLObjectType({
+  name: "Foo",
+  fields: () => ({
+    id: {
+      type: new graphql.GraphQLNonNull(graphql.GraphQLID),
+    },
+    bar: {
+      // Allows referring to types by their string names
+      type: TypeRegistry.getType("Bar"),
+    },
+  }),
+})
+```
+
+After all types are collected into type registry, it can be converted to a normal GraphQL schema. Other common features include being able to generate types like input objects and filter from the types held in type registry.
+
+We didn't want to implement type registry and all the related things ourselves. Thankfully, there is a library just for that - [graphql-compose](https://graphql-compose.github.io/). We opted to use it and it saved us lots of time. I really recommend this library to anyone, especially if you plan to generate the types.
+
+```js
+// global schema composer
+const { SchemaComposer } = require("graphql-compose")
+
+const Foo = SchemaComposer.TypeComposer.create(({
+  name: "Foo",
+  fields: () => ({
+    id: {
+      // types can be strings
+      type: 'ID!'
+    },
+    bar: {
+      // Allows referring to types by their string names
+      type: 'Bar'
+    },
+  }),
+})
+```
+
+The final schema pipeline that we implemented works like this:
+
+1. We collect all types that are created with `createTypes` and add them to compose type registry (called _Schema Composer_).
+2. We go through all the collected nodes and we infer types for them
+3. We merge user defined types with inferred types and add them to the composer
+4. We add default resolvers for type fields, such as for `File` and `Date` fields
+5. `setFieldsOnNodeType` is called and those fields are added to the types.
+6. We create derived input objects, such as filter and sort and then create pagination types such as Connections
+7. Root level resolvers are created for all node types
+8. `createResolvers` api is called and resulting resolvers are added to the schema
+9. We generate the schema
+
+You can see the TODO LINK `packages/gatsby/schema/` folder in the schema= refactoring branch to learn more about the code.
+
```

Comment:
Do we want to mention the `nodes` shortcut for `edges { node }` as "Bonus material"?